underfs/cephfs/src/main/java/alluxio/underfs/cephfs/CephFSUnderFileSystem.java

```diff
@@ -0,0 +1,792 @@
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_MDS_NAMESPACE, namespace);
+    if (namespace != null && !namespace.isEmpty()) {
+      mount.conf_set("client_mds_namespace", namespace);
+    }
+
+    /*
+     * Set uid/gid to mount
+     */
+    String uid = conf.get(PropertyKey.UNDERFS_CEPHFS_MOUNT_UID);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_MOUNT_UID, uid);
+    if (uid != null && !uid.isEmpty()) {
+      mount.conf_set("client_mount_uid", uid);
+    }
+
+    String gid = conf.get(PropertyKey.UNDERFS_CEPHFS_MOUNT_GID);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_MOUNT_GID, gid);
+    if (gid != null && !gid.isEmpty()) {
+      mount.conf_set("client_mount_gid", gid);
+    }
+
+    /*
+     * Actually mount the file system
+     */
+    String root = conf.get(PropertyKey.UNDERFS_CEPHFS_MOUNT_POINT);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_MOUNT_POINT, root);
+    mount.mount(root);
+
+    /*
+     * Allow reads from replica objects?
+     */
+    String localizeReads = conf.get(PropertyKey.UNDERFS_CEPHFS_LOCALIZE_READS);
+    LOG.info("CephFS config: {} = {}", PropertyKey.UNDERFS_CEPHFS_LOCALIZE_READS, localizeReads);
+    if (localizeReads != null && !localizeReads.isEmpty()) {
+      boolean bLocalize = Boolean.parseBoolean(localizeReads);
+      mount.localize_reads(bLocalize);
+    }
+
+    return new CephFSUnderFileSystem(ufsUri, mount, conf);
+  }
+
+  /**
+   * Constructs a new CephFS {@link UnderFileSystem}.
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param mount CephMount instance
+   * @param conf the configuration for this UFS
+   */
+  public CephFSUnderFileSystem(AlluxioURI ufsUri, CephMount mount,
+      UnderFileSystemConfiguration conf) {
+    super(ufsUri, conf);
+    mMount = mount;
+  }
+
+  @Override
+  public String getUnderFSType() {
+    return "cephfs";
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (null != mMount) {
+      mMount.unmount();
+    }
+    mMount = null;
+  }
+
+  @Override
+  public OutputStream create(String path, CreateOptions options) throws IOException {
+    if (!options.isEnsureAtomic()) {
+      return createDirect(path, options);
+    }
+    return new AtomicFileOutputStream(path, this, options);
+  }
+
+  @Override
+  public void cleanup() throws IOException {
+  }
+
+  @Override
+  public OutputStream createNonexistingFile(String path) throws IOException {
+    return create(path);
+  }
+
+  @Override
+  public OutputStream createNonexistingFile(String path, CreateOptions options) throws IOException {
+    return create(path, options);
+  }
+
+  @Override
+  public OutputStream createDirect(String path, CreateOptions options) throws IOException {
+    path = stripPath(path);
+    String parentPath;
+    try {
+      parentPath = PathUtils.getParent(path);
+    } catch (InvalidPathException e) {
+      throw new IOException("Invalid path");
```

Comment:
```suggestion
      throw new IOException("Invalid path " + path, e);
```
to help trouble-shooting